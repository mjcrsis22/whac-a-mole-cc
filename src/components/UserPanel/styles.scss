.user-panel {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem 1rem;
  background-image: url(/backgrounds/zig-zag-crimson.svg);
  background-size: contain;
  background-color: crimson;
  color: white;

  .user-info-container {
    display: flex;
    flex-direction: row;
    align-items: center;

    .exit-button {
      padding: 0.2rem 0.4rem;
      margin-left: 0.7rem;
      color: crimson;
      font-weight: bold;

      &:hover {
        border-color: crimson;
      }
    }
  }
}
