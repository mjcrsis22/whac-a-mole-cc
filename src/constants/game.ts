export enum GAME_LEVEL {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HARD = 'HARD',
}

export interface GameSetting {
  id: GAME_LEVEL;
  spotRate: number;
  points: number;
}

export const GAME_SETTINGS_BY_LEVEL: {
  [key in keyof typeof GAME_LEVEL]: GameSetting;
} = {
  [GAME_LEVEL.LOW]: {
    id: GAME_LEVEL.LOW,
    spotRate: 1000,
    points: 10,
  },
  [GAME_LEVEL.MEDIUM]: {
    id: GAME_LEVEL.MEDIUM,
    spotRate: 750,
    points: 20,
  },
  [GAME_LEVEL.HARD]: {
    id: GAME_LEVEL.HARD,
    spotRate: 500,
    points: 30,
  },
};
