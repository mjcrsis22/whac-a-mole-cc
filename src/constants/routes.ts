export enum APP_ROUTES {
  HOME = '/home',
  GAME = '/game',
}
