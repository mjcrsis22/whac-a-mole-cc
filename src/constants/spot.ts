export enum SPOT_STATES {
  inactive,
  active,
}

export const SPOT_POSITIONS = [
  { left: '5%', top: '0%' },
  { left: '5%', top: '72%' },
  { left: '30%', top: '15%' },
  { left: '30%', top: '60%' },
  { left: '45%', top: '40%' },
  { left: '60%', top: '15%' },
  { left: '60%', top: '60%' },
  { left: '85%', top: '0%' },
  { left: '85%', top: '72%' },
];
