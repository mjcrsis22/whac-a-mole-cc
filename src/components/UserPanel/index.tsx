import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import UserScoring from '../UserScoring';
import {
  removeStoredData,
  getStoredData,
  STORE_KEYS,
} from '../../utils/storage';
import { APP_ROUTES } from '../../constants/routes';
import './styles.scss';

function UserPanel() {
  const navigate = useNavigate();
  const [username, setUsername] = useState('');

  const handleExit = () => {
    removeStoredData(STORE_KEYS.USERNAME);
    navigate(APP_ROUTES.HOME);
  };

  useEffect(() => {
    const item = getStoredData(STORE_KEYS.USERNAME);
    if (item) {
      setUsername(item);
    } else {
      navigate(APP_ROUTES.HOME);
    }
  }, [navigate]);

  return (
    <div className="user-panel">
      <UserScoring />
      <div className="user-info-container">
        <span className="username">{username}</span>
        <button className="button exit-button" onClick={handleExit} title='Exit'>
          X
        </button>
      </div>
    </div>
  );
}

export default UserPanel;
