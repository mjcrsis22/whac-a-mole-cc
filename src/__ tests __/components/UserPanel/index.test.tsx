import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  STORE_KEYS,
  getStoredData,
  removeStoredData,
} from '../../../utils/storage';
import { APP_ROUTES } from '../../../constants/routes';
import UserPanel from '../../../components/UserPanel';

const mockedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(() => mockedNavigate),
}));

jest.mock('../../../utils/storage', () => ({
  ...jest.requireActual('../../../utils/storage'),
  getStoredData: jest.fn(),
  removeStoredData: jest.fn(),
}));

const mockedGetStoredData = getStoredData as jest.MockedFn<
  typeof getStoredData
>;

describe('UserPanel component', () => {
  beforeEach(() => {
    mockedGetStoredData.mockReturnValue('MiNombre');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Renders correctly', () => {
    const { container } = render(<UserPanel />);
    expect(container).toMatchSnapshot();
  });

  test('Renders correctly', () => {
    mockedGetStoredData.mockReturnValue(null);
    const { container } = render(<UserPanel />);
    expect(container).toMatchSnapshot();
    expect(mockedNavigate).toHaveBeenCalledWith(APP_ROUTES.HOME);
  });

  test('Calls removeStoredData and navigate button click', async () => {
    const result = render(<UserPanel />);

    const button = result.getByRole('button');
    await userEvent.click(button);

    expect(removeStoredData).toHaveBeenCalledWith(STORE_KEYS.USERNAME);
    expect(mockedNavigate).toHaveBeenCalledWith(APP_ROUTES.HOME);
  });
});
