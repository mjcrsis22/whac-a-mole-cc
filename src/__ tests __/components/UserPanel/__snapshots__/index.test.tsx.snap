// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPanel component Renders correctly 1`] = `
<div>
  <div
    class="user-panel"
  >
    Score: 
    0
    <div
      class="user-info-container"
    >
      <span
        class="username"
      >
        MiNombre
      </span>
      <button
        class="button exit-button"
        title="Exit"
      >
        X
      </button>
    </div>
  </div>
</div>
`;

exports[`UserPanel component Renders correctly 2`] = `
<div>
  <div
    class="user-panel"
  >
    Score: 
    0
    <div
      class="user-info-container"
    >
      <span
        class="username"
      />
      <button
        class="button exit-button"
        title="Exit"
      >
        X
      </button>
    </div>
  </div>
</div>
`;
